### PR TITLE
Create CLI Verify subcommand which validates given Config toml file before Deployment

### PR DESCRIPTION
**Pull request by atultherajput**
### :nut_and_bolt: Description: What code changed, and why?

Create CLI Verify subcommand which validates the given Config toml file before Deployment.

### :chains: Related Resources
https://chefio.atlassian.net/browse/CHEF-2053
### :white_check_mark: Checklist

**All PRs** must tick these:

- [ ] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [ ] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [ ] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [ ] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [ ] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [ ] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [ ] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [ ] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### :camera: Screenshots, if applicable
https://progresssoftware.sharepoint.com/:v:/s/ChefCoreC/EbT_ck2d6XJBkgGb6r-Eh_ABvRHq3-iEFn1YcBq4KI92-A?e=pxn6C7

Screenshot 2023-06-19 at 1 33 49 PM

Screenshot 2023-06-19 at 3 25 07 PM

### CODE REVIEW

File: lib/sshutils/sshutils.go [hunk 1]
```diff
@@ -171,6 +181,110 @@ func (s *SSHUtilImpl) Execute(sshConfig SSHConfig, cmd string) (string, error) {
...
+	for _, hostIP := range hostIPs {
+		sshConfig.HostIP = hostIP
+		s.logger.Debugln("Connecting to the node : " + hostIP)
+
+		go func(sshConfig SSHConfig, cmd string, resultChan chan Result) {
+			rc := Result{sshConfig.HostIP, "", nil}
+
+			output, err := s.Execute(sshConfig, cmd)
+			if err != nil {
+				rc.Error = err
```
Comment (vipin230) [inline lib/sshutils/sshutils.go:196]: return output also

File: components/automate-cli/pkg/verifysystemdcreate/systemdcreatesvc.go [hunk 2]
```diff
@@ -121,6 +121,22 @@ func (css *CreateSystemdService) createSystemdServiceFile() error {
 	return nil
 }
 
+// Check if status is active and then stop the service.
+func (css *CreateSystemdService) stopIfExistSystemdService() error {
```
Comment (bvtejaswi) [inline components/automate-cli/pkg/verifysystemdcreate/systemdcreatesvc.go:125]:
```suggestion
func (css *CreateSystemdService) stopIfSystemdServiceIsActive() error {
```

File: components/automate-cli/cmd/chef-automate/verify.go [hunk 3]
```diff
@@ -204,126 +185,244 @@ func (v *verifySystemdCreateFlow) runVerifySystemdCreateCmd(cmd *cobra.Command,
...
+	// Doing batch-check API call for remote nodes
+	batchCheckRemoteReq := models.BatchCheckRequest{
+		Checks: constants.GetRemoteChecks(),
+		Config: batchCheckConfig,
 	}
-	if err := v.Verification.VerifyStandaloneDeployment(configPath); err != nil {
-		return status.Annotate(err, status.ConfigError)
+
+	v.Writer.Println("Doing batch-check API call for remote nodes")
+	batchCheckRemoteRes, err := v.Client.MakeRequest("POST", fmt.Sprintf("http://localhost:%s/api/v1/checks/batch-checks", getPort()), batchCheckRemoteReq)
```
Comment (shaik80) [inline components/automate-cli/cmd/chef-automate/verify.go:375]: get endpoint from const, as it is defined in 2 places

**Review by shaik80 — CHANGES_REQUESTED** [review]

File: lib/sshutils/sshutils.go [hunk 4]
```diff
@@ -171,6 +181,110 @@ func (s *SSHUtilImpl) Execute(sshConfig SSHConfig, cmd string) (string, error) {
...
+		}(sshConfig, cmd, resultChan)
+	}
+
+	var results []Result
+
+	for i := 0; i < len(hostIPs); i++ {
+		result := <-resultChan
+
+		if result.Error != nil {
+			s.logger.Errorf("Remote executation failed on node : %s with error: %v\n", result.HostIP, result.Error)
```
Comment (bvtejaswi) [inline lib/sshutils/sshutils.go:218]:
```suggestion
			s.logger.Errorf("Remote execution failed on node : %s with error: %v\n", result.HostIP, result.Error)
```

File: lib/sshutils/sshutils.go [hunk 5]
```diff
@@ -171,6 +181,110 @@ func (s *SSHUtilImpl) Execute(sshConfig SSHConfig, cmd string) (string, error) {
...
+		} else {
+			s.logger.Debugf("Remote executation is completed. Output for Host IP %s : \n%s\n", result.HostIP, result.Output)
```
Comment (bvtejaswi) [inline lib/sshutils/sshutils.go:220]:
```suggestion
			s.logger.Debugf("Remote execution is completed. Output for Host IP %s : \n%s\n", result.HostIP, result.Output)
```

File: components/automate-cli/cmd/chef-automate/verify.go [hunk 6]
```diff
@@ -204,126 +185,244 @@ func (v *verifySystemdCreateFlow) runVerifySystemdCreateCmd(cmd *cobra.Command,
...
 
 	return nil
 }
 
-func (v *verifyCmdFlow) verifyHaAWSDeploy(configPath string) error {
-	if !v.ManagedServicesOn {
-		if v.A2HARBFileExist {
-			if err := v.Verification.VerifyHAAWSDeployment(configPath); err != nil {
-				return status.Annotate(err, status.ConfigError)
+func (v *verifyCmdFlow) checkAutomateVerifyServiceForBastion(batchCheckConfig models.Config) error {
```
Comment (vipin230) [inline components/automate-cli/cmd/chef-automate/verify.go:248]: break into multiple functions one for check and other for api call

File: lib/sshutils/sshutils.go (hunk 1 above)
Comment (atultherajput) [reply lib/sshutils/sshutils.go:205]: done

File: components/automate-cli/cmd/chef-automate/verify.go (hunk 3 above)
Comment (atultherajput) [reply components/automate-cli/cmd/chef-automate/verify.go]: done

File: lib/sshutils/sshutils.go (hunk 4 above)
Comment (atultherajput) [reply lib/sshutils/sshutils.go]: done

File: lib/sshutils/sshutils.go (hunk 5 above)
Comment (atultherajput) [reply lib/sshutils/sshutils.go]: done

File: components/automate-cli/cmd/chef-automate/verify.go [hunk 7]
```diff
@@ -204,126 +185,244 @@ func (v *verifySystemdCreateFlow) runVerifySystemdCreateCmd(cmd *cobra.Command,
...
+
+	return nil
 }
 
-func (v *verifyCmdFlow) verifyHaAWSManagedDeploy(configPath string) error {
-	if v.ManagedServicesOn {
-		if v.A2HARBFileExist {
-			if err := v.Verification.VerifyHAAWSManagedDeployment(configPath); err != nil {
-				return status.Annotate(err, status.ConfigError)
+func (v *verifyCmdFlow) checkAutomateVerifyServiceForRemote(batchCheckConfig models.Config) error {
```
[the remark this answers is not in the document]
Comment (atultherajput) [reply components/automate-cli/cmd/chef-automate/verify.go]: done

File: components/automate-cli/cmd/chef-automate/verify.go (hunk 6 above)
Comment (atultherajput) [reply components/automate-cli/cmd/chef-automate/verify.go]: done

File: components/automate-cli/cmd/chef-automate/verify.go [hunk 8]
```diff
@@ -204,126 +197,281 @@ func (v *verifySystemdCreateFlow) runVerifySystemdCreateCmd(cmd *cobra.Command,
...
+	err := v.Config.ParseAndVerify(configPath)
+	if err != nil {
 		return err
-	case flagsObj.haOnPremCustManagedDeploy:
-		err := v.verifyHaOnPremCustManagedDeploy(configPath)
+	}
+
+	// Get config required for batch-check API call
+	batchCheckConfig := &models.Config{}
+	batchCheckConfig.PopulateWith(v.Config)
```
Comment (Dmaddu) [inline components/automate-cli/cmd/chef-automate/verify.go:243]: This returns error, please handle the error

File: components/automate-cli/pkg/verifyserver/server/api/v1/status.go [hunk 9]
```diff
@@ -4,21 +4,24 @@ import (
...
 func (h *Handler) GetStatus(c *fiber.Ctx) error {
 	services, err := h.StatusService.GetServices()
 	if err != nil {
 		return c.JSON(response.BuildSuccessResponse(&models.StatusDetails{
-			Status:   constants.OK,
-			Services: &[]models.ServiceDetails{},
-			Error:    err.Error(),
+			Status:     constants.OK,
+			Services:   &[]models.ServiceDetails{},
+			CliVersion: version.BuildTime,
```
Comment (Dmaddu) [inline components/automate-cli/pkg/verifyserver/server/api/v1/status.go:17]: Please add a demo video, showing this response from the status API

File: components/automate-cli/cmd/chef-automate/verify.go [hunk 10]
```diff
@@ -1,42 +1,71 @@
 package main
 
 import (
+	"encoding/json"
```
Comment (Dmaddu) [inline components/automate-cli/cmd/chef-automate/verify.go:4]: The test coverage is less, Please improve the coverage
image

**Review by Dmaddu — CHANGES_REQUESTED** [review]
To unblock others for now, please add a demo video, For remaining changes please create another PR

File: components/automate-cli/cmd/chef-automate/verify.go (hunk 8 above)
Comment (atultherajput) [reply components/automate-cli/cmd/chef-automate/verify.go]: done

File: components/automate-cli/cmd/chef-automate/verify.go (hunk 10 above)
Comment (atultherajput) [reply components/automate-cli/cmd/chef-automate/verify.go:4]: done

File: components/automate-cli/cmd/chef-automate/verify.go [hunk 11]
```diff
@@ -204,126 +197,284 @@ func (v *verifySystemdCreateFlow) runVerifySystemdCreateCmd(cmd *cobra.Command,
...
+		v.Writer.Println(fmt.Sprintf("error while checking automate-verify service on bastion: %v\n", err))
+		v.Writer.Println("Adding automate-verify service to systemd and starting the service")
+
+		// create systemd service with current CLI binary
+		err = v.createSystemdOnBastion()
+		if err != nil {
+			return err
+		}
+
+		v.Writer.Println("Adding automate-verify service to systemd and starting the service completed")
```
[the remark this answers is not in the document]
Comment (atultherajput) [reply components/automate-cli/cmd/chef-automate/verify.go]: done

File: components/automate-cli/pkg/verifyserver/server/api/v1/status.go (hunk 9 above)
Comment (atultherajput) [reply components/automate-cli/pkg/verifyserver/server/api/v1/status.go:17]: done

File: components/automate-cli/cmd/chef-automate/verify.go [hunk 12]
```diff
@@ -204,126 +197,286 @@ func (v *verifySystemdCreateFlow) runVerifySystemdCreateCmd(cmd *cobra.Command,
...
+			v.Writer.Errorf("Remote execution of cmd '%s' automate-verify CLI failed on node : %s with error: %v\n", cmd, result.HostIP, result.Error)
+			isError = true
+		}
 	}
+	if isError {
+		return fmt.Errorf("remote execution failed")
+	}
+	// Wait for given seconds for the service to start
+	time.Sleep(SLEEP_DURATION)
+	v.Writer.Println("Creating automate-verify service on remote nodes completed")
```
Comment (Sunanda-Boorla) [inline components/automate-cli/cmd/chef-automate/verify.go:427]: Can we have "Created automate-verify service on remote nodes"

**Review by Dmaddu — APPROVED** [review]

**Review by vviveksharma — APPROVED** [review]

File: components/automate-cli/cmd/chef-automate/verify.go (hunk 12 above)
Comment (atultherajput) [reply components/automate-cli/cmd/chef-automate/verify.go]: done

File: components/automate-cli/cmd/chef-automate/verify.go [hunk 13]
```diff
@@ -204,126 +197,286 @@ func (v *verifySystemdCreateFlow) runVerifySystemdCreateCmd(cmd *cobra.Command,
...
+
+	var resultBytes []byte
+	if v.prettyPrint {
+		resultBytes, err = json.MarshalIndent(responseBodyStruct.Result, "", "    ")
+	} else {
+		resultBytes, err = json.Marshal(responseBodyStruct.Result)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal Result field: %v", err)
```
[the remark this answers is not in the document]
Comment (atultherajput) [reply components/automate-cli/cmd/chef-automate/verify.go]: done

**Review by bvtejaswi — APPROVED** [review]

**Review by rishabhjhs — APPROVED** [review]

**Review by punitmundra — APPROVED** [review]

**Pull request merged**